src/embind/embind.js

```diff
@@ -1758,29 +1758,33 @@ var LibraryEmbind = {
     // at run-time, not build-time.
     finalizationRegistry = new FinalizationRegistry(function (info) {
 #if ASSERTIONS
-      console.warn(info.leakWarning.stack.replace(/^Error: /, ''));
+        console.warn(info.leakWarning.stack.replace(/^Error: /, ''));
 #endif
-      releaseClassHandle(info.$$);
+        releaseClassHandle(info.$$);
     });
     attachFinalizer = function(handle) {
-      var $$ = handle.$$;
-      var info = { $$: $$ };
+        var $$ = handle.$$;
+        var hasSmartPtr = !!$$.smartPtr;
```

Comment:
Interesting, didn't realise it would be so simple. Just in case, did you check that it indeed works for all supported smart pointers and doesn't free the raw ones?

Reply:
@RReverser I checked that it frees shared_ptr and unique_ptr, and doesn't free the raw pointer. Are there more?

Reply:
I think that sounds good! Maybe @kripken or @sbc100 have other ideas?

Reply:
Hmm, no additional ideas - that sounds good.

What kind of tests do we have here?

Reply:
A possible test could be: start Node.js with `--expose-gc`, create an object, let it go out of scope, force a GC pass with `global.gc();` and check the output for the leak warning, or for something that the destructor prints.

Reply:
That test sounds good to me! And we could verify that the same does not happen with a raw pointer.

Reply:
As mentioned in the past PR, I've tried to make tests in that way, but they've been flaky, and the spec doesn't guarantee when finalizers callback will actually run relative to the GC pass (it can happen immediately, or a bit later, or never).

You can try adding such test, and perhaps adding some `setTimeout` to wait a bit for the callback, just saying that it's likely to still be flaky in the future.

Reply:
@kripken Waiting 10s after the gc call seems to run the destructors consistently. Do you think I should add that test to this PR, or would you rather avoid the risk of a future Node version breaking it?

(Also, the failing tests in this PR right now seem to be not my fault: before the last commit they were passing, and I changed nothing that would be relevant to the failures. Something's up with the CI environment, something to do with a too new binaryen version?)

Reply:
10s sounds like a lot even for GC... Did you try smaller durations?

Reply:
@RReverser Waiting 8s after `global.gc()` is not enough for the finalization to happen. 9s usually is.

Reply:
You can try going the other way around like I did in previous PR [here](https://github.com/emscripten-core/emscripten/pull/15327#issuecomment-946828761) - that is, waiting before calling `gc()`. That seemed to work even with 100ms. One way or another it will be flaky at some point, but if we want to go with tests anyway, that might help reduce the duration.

Reply:
@RReverser Interesting! If I `setTimeout` before `gc`, even a timeout value of 0 is sufficient for the finalization to reliably trigger.

Reply:
I would expect `gc()` to synchronously GC and collect everything possible. It is odd that a `setTimeout` is needed for that to happen... but if it works, sgtm...

Though I do remember @RReverser that as you said we did look into this before and were discouraged.

Reply:
Yeah, the difference is between _collection_ and _finalization_ - the latter is not guaranteed to happen right after GC (or at all).

It does seem to work with `setTimeout` for now, and I guess it's better than nothing for tests, so we can merge those tests as-is and remove or adjust in future if the behaviour changes on the Node side.